src/libs/sys-util-rs/src/mount.rs

```diff
@@ -0,0 +1,1073 @@
+    fn mount<P: AsRef<Path>>(&self, target: P) -> Result<()>;
+}
+
+impl Mounter for kata_types::mount::Mount {
+    // This function is modelled after
+    // [Mount::Mount()](https://github.com/containerd/containerd/blob/main/mount/mount_linux.go)
+    // from [Containerd](https://github.com/containerd/containerd) project.
+    fn mount<P: AsRef<Path>>(&self, target: P) -> Result<()> {
+        fail::fail_point!("Mount::mount", |_| {
+            Err(Error::FailureInject(
+                "Mount::mount() fail point injection".to_string(),
+            ))
+        });
+
+        let target = target.as_ref().to_path_buf();
+        let (chdir, m_opts) =
+            // Follow the same algorithm as Containerd: reserve 512 bytes to avoid hitting one page
+            // limit of mounting argument buffer.
+            if self.fs_type == "overlay" && self.option_size() >= *MAX_MOUNT_PARAM_SIZE - 512 {
+                info!(
+                    sl!(),
+                    "overlay mount option too long, maybe failed to mount"
+                );
+                let (chdir, options) = compact_lowerdir_option(&self.options);
+                (chdir, parse_mount_option(&options)?)
+            } else {
+                (PathBuf::new(), parse_mount_option(&self.options)?)
+            };
+
+        // Ensure propagation type change flags aren't included in other calls.
+        let propagation_types =
+            MsFlags::MS_SHARED | MsFlags::MS_PRIVATE | MsFlags::MS_SLAVE | MsFlags::MS_UNBINDABLE;
+        let o_flag = m_opts.flag & (!propagation_types);
+
+        // - Normal mount without MS_REMOUNT flag
+        // - In the case of remounting with changed data (data != ""), need to call mount
+        if (m_opts.flag & MsFlags::MS_REMOUNT) == MsFlags::empty() || !m_opts.data.is_empty() {
+            mount_at(
+                chdir,
+                &self.source,
+                target.clone(),
+                &self.fs_type,
+                o_flag,
+                &m_opts.data,
+            )?;
+        }
+
+        // Change mount propagation type.
+        if (m_opts.flag & propagation_types) != MsFlags::empty() {
+            let propagation_flag = propagation_types | MsFlags::MS_REC | MsFlags::MS_SILENT;
+            mount(
+                Some(""),
+                &target,
+                Some(""),
+                m_opts.flag & propagation_flag,
+                Some(""),
+            )
+            .map_err(|e| Error::Mount(PathBuf::new(), target.clone(), e))?;
+        }
+
+        // Bind mount readonly.
+        let bro_flag = MsFlags::MS_BIND | MsFlags::MS_RDONLY;
+        if (o_flag & bro_flag) == bro_flag {
+            mount(
+                Some(""),
+                &target,
+                Some(""),
+                o_flag & MsFlags::MS_REMOUNT,
+                Some(""),
+            )
+            .map_err(|e| Error::BindMount(PathBuf::new(), target, e))?;
+        }
```

Comment:
They looks similar, but are really different:) And they have different purpose too.
`bind_mount()` acts a generic utility function, the above code is for `impl Mounter for kata_types::mount::Mount` which is used to mount a kata-types::Mount.

Reply:
Can't you have both paths call a common helper though?  It might need some extra parameter, but you should still be able to avoid duplicating the whole thing.

Reply:
fixed